src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php

```diff
@@ -123,38 +131,60 @@ private function registerHandlers(ContainerBuilder $container)
                         $definitions[$serviceId = '.messenger.method_on_object_wrapper.'.ContainerBuilder::hash($messageClass.':'.$messagePriority.':'.$serviceId.':'.$method)] = $wrapperDefinition;
                     }
 
-                    $handlersByMessage[$messageClass][$messagePriority][] = new Reference($serviceId);
+                    foreach ($handlerBuses as $handlerBus) {
+                        $handlersByBusAndMessage[$handlerBus][$messageClass][$messagePriority][] = $serviceId;
+                    }
                 }
             }
         }
 
-        foreach ($handlersByMessage as $message => $handlers) {
-            krsort($handlersByMessage[$message]);
-            $handlersByMessage[$message] = array_merge(...$handlersByMessage[$message]);
+        foreach ($handlersByBusAndMessage as $bus => $handlersByMessage) {
+            foreach ($handlersByMessage as $message => $handlersByPriority) {
+                krsort($handlersByPriority);
+                $handlersByBusAndMessage[$bus][$message] = array_unique(array_merge(...$handlersByPriority));
+            }
         }
 
-        $handlersLocatorMapping = array();
-        foreach ($handlersByMessage as $message => $handlers) {
-            if (1 === \count($handlers)) {
-                $handlersLocatorMapping['handler.'.$message] = current($handlers);
-            } else {
-                $d = new Definition(ChainHandler::class, array($handlers));
-                $d->setPrivate(true);
-                $serviceId = hash('sha1', $message);
-                $definitions[$serviceId] = $d;
-                $handlersLocatorMapping['handler.'.$message] = new Reference($serviceId);
+        $handlersLocatorMappingByBus = array();
+        foreach ($buses as $bus) {
+            $handlersLocatorMappingByBus[$bus] = array();
```

Comment:
```php
$handlersLocatorMappingByBus = array_combine($busIds, array_fill(0, \count($busIds), array());
```

Not sure if it's more efficient though 😄

Reply:
Yes, I didn’t do it because I think it’s less explicit 😅